requirements.txt

```diff
@@ -30,14 +30,14 @@ rouge-score>=0.1.2,<0.2.0
 sentencepiece>=0.2.0,<0.3.0
 tokenizers>=0.15.2,<0.16.0
 toml>=0.10.2,<0.11.0
-# Habana installer has 2.2.0a0+git8964477 with oneMKL
-torch>=2.2.0a0,<3.0.0 ; python_version == '3.10'
-torch>=2.2.1,<3.0.0 ; python_version != '3.10'
+torch>=2.3.0,<3.0.0
 tqdm>=4.66.2,<5.0.0
-transformers>=4.30.0,<=4.38.2
+transformers==4.38.1
 trl>=0.7.11,<0.8.0
 wandb>=0.16.4,<0.17.0
 langchain-text-splitters
 # the below library should NOT be imported into any python files
 # it is for CLI usage ONLY
 yamllint>=1.35.1,<1.36.0
+bitsandbytes
+accelerate
```

Comment:
I see there is a `requirements-hpu.txt` I don't think we should have niche stack's in our main requirements.txt. 

Would having that other requirements.txt contain all gaudi reqs be ok?

Reply:
@tiran ok, here are the requirements we have overlapping:

Gaudi: Python 3.10, torch 2.2.0a
MPS and all other Devices: 3.10+, torch 2.3.0+

Can you please advise the best way to merge these two requirements? 

I was thinking `pip install  -r requirements-hpu.txt -e .` would allow us to override the torch version in requirements.txt. 

One thing I was thinking of would be to have `make hpu` or `make mps` or `make cuda` targets that also build ilab to be able to handle these backends?

I am open to whatever way is best! Though, I don't think having this stuff in the base requirements file makes sense since conflicts like this are bound to happen a lot.

Reply:
Optional dependencies cannot introduce conflicting dependencies. They are limited to additional dependencies or more restrictive dependencies. Pip flat-out refuses to resolve conflicts:

```
ERROR: Cannot install instructlab[hpu]==0.14.1.dev183+gdeee169.d20240514 because these package versions have conflicting dependencies.

The conflict is caused by:
    instructlab[hpu] 0.14.1.dev183+gdeee169.d20240514 depends on torch<3.0.0 and >=2.3.0
    instructlab[hpu] 0.14.1.dev183+gdeee169.d20240514 depends on torch==2.2.0a0; extra == "hpu"
```

Why do we require PyTorch 2.3.0 anyway? Is there a specific feature that is required?

Reply:
mps backend is broken below 2.3. I can test it out to give you the specific reason. However, I don't think it can compile with MPS below 2.3